### PR TITLE
Update `config@3.3.7` (from `3.3.3`)

### PR DESCRIPTION
**Pull request by tomalec**
:warning: this PR is based on https://github.com/woocommerce/woocommerce/pull/33827 - it does not have to be connected, but it's easier to guide to the actual use- and test case I have.

### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

[Update `config@3.3.7` (from `3.3.3`)](https://github.com/woocommerce/woocommerce/commit/10301f5179aac79a8dc3ace8b71414b698ec7f41) 

to avoid `ReferenceError: node_env_var_name is not defined` when external extension imports and transforms the `admin-e2e-tests` directly from `node_modules`.

Include https://github.com/node-config/node-config/pull/642

#### Bug repro
1. Clone an extension that uses it, like https://github.com/woocommerce/google-listings-and-ads/tree/b14771982bb0c84d6252b27f36684114da06b85a
2. run `npx wc-e2e docker:up && npx wc-e2e test:e2e`

#### fix

1. Clone an extension that uses it, like https://github.com/woocommerce/google-listings-and-ads/tree/b14771982bb0c84d6252b27f36684114da06b85a
2. ~Install this version of e2e-*~
   ```
   cd node_modules/@woocommerce/e2e-environment/
   npm i --save jest-puppeteer@^5.0.4
   npm i --save config@^3.3.7
   cd ../e2e-utils/
   npm i --save config@^3.3.7
   cd ../../../
   ```
3. run `npx wc-e2e docker:up && npx wc-e2e test:e2e`

### Other information:

I was not able to run build and fully test  locally due to https://github.com/woocommerce/woocommerce/issues/33684

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?
   ```
   None of the selected packages has a "changelog" script
   ```
   So I added the changelog entry manually to `CHANGELOG.md`

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Comment by Luc45** [comment]
This looks like a pretty straightforward change, but e2e tests are all failing on CI in this PR.

I have re-run them just to be sure, and the failures are consistent. I think e2e-environment might be unable to load the config using the updated version for some reason.

**Comment by tomalec** [comment]
> I think e2e-environment might be unable to load the config using the updated version for some reason.

Do you have any hint how to debug it locally?
As mentioned in https://github.com/woocommerce/woocommerce/pull/33852 I cannot even run build command locally.

**Comment by tomalec** [comment]
I also guess it may fail due to lack of appPath to resolve config paths, due to https://github.com/woocommerce/woocommerce/issues/33684

**Comment by Luc45** [comment]
> I also guess it may fail due to lack of appPath to resolve config paths, due to https://github.com/woocommerce/woocommerce/issues/33684

I don't think so, because other PRs that are up-to-date with `trunk` are passing: https://github.com/woocommerce/woocommerce/pull/33905

This seems to have something to do with the `config` package. I'll take a closer look into it. However, we are in the process of deprecating `e2e-environment`, so I strongly suggest using the new setup based on Playwright + wp-env.

We are just waiting for this to be merged: https://github.com/woocommerce/woocommerce/pull/33850

Then you can follow this documentation (feedback on documentation is highly appreciated): https://github.com/woocommerce/woocommerce/blob/db50c3249720711f61f7bae29b8a7b5ffba53ba2/plugins/woocommerce/e2e/README.md

**Comment by tomalec** [comment]
> I don't think so, because other PRs that are up-to-date with `trunk` are passing: https://github.com/woocommerce/woocommerce/pull/33905

:ok_hand: 

---

> However, we are in the process of deprecating e2e-environment, so I strongly suggest using the new setup based on Playwright + wp-env.

Cool, thanks. That's something new to me.

> Then you can follow this documentation (feedback on documentation is highly appreciated): [db50c32/plugins/woocommerce/e2e/README.md](https://github.com/woocommerce/woocommerce/blob/db50c3249720711f61f7bae29b8a7b5ffba53ba2/plugins/woocommerce/e2e/README.md)

This doc seem to explain how to test woocommerce itself. What I'm trying to do is to test [my Woo extension](https://github.com/woocommerce/google-listings-and-ads)  with `admin-e2e-tests` suite. Do you have some guides on that specifically? as [its README](https://github.com/woocommerce/woocommerce/tree/db50c3249720711f61f7bae29b8a7b5ffba53ba2/packages/js/admin-e2e-tests) still suggests the use of `e2e-environment`

**Comment by alopezari** [comment]
Hi @tomalec!

First of all, thank you very much for creating this PR.

We might need to update that documentation since the tests exposed [at the bottom of the page](https://github.com/woocommerce/woocommerce/tree/db50c3249720711f61f7bae29b8a7b5ffba53ba2/packages/js/admin-e2e-tests) are already covered by the new setup based on Playwright that @Luc45 mentioned.

If you check the [table from the Admin E2E tests documentation](https://github.com/woocommerce/woocommerce/tree/db50c3249720711f61f7bae29b8a7b5ffba53ba2/packages/js/admin-e2e-tests), you will see that they are aligned with the [activate-and-setup](https://github.com/woocommerce/woocommerce/tree/db50c3249720711f61f7bae29b8a7b5ffba53ba2/plugins/woocommerce/e2e/tests/activate-and-setup), [admin-analytics](https://github.com/woocommerce/woocommerce/tree/db50c3249720711f61f7bae29b8a7b5ffba53ba2/plugins/woocommerce/e2e/tests/admin-analytics), [admin-marketing](https://github.com/woocommerce/woocommerce/tree/db50c3249720711f61f7bae29b8a7b5ffba53ba2/plugins/woocommerce/e2e/tests/admin-marketing), [admin-tasks](https://github.com/woocommerce/woocommerce/tree/db50c3249720711f61f7bae29b8a7b5ffba53ba2/plugins/woocommerce/e2e/tests/admin-tasks) and [basic.spec.js](https://github.com/woocommerce/woocommerce/blob/db50c3249720711f61f7bae29b8a7b5ffba53ba2/plugins/woocommerce/e2e/tests/basic.spec.js) Playwright tests.

So yes, you can rely on these Playwright specs to test this PR :) 

You can find [here](https://github.com/woocommerce/woocommerce/blob/db50c3249720711f61f7bae29b8a7b5ffba53ba2/plugins/woocommerce/e2e/README.md#introduction) more information about how to run these tests. In concrete, you can use this example to run individually the specs mentioned above, which covers the same as the previous Admin E2E tests:
```
cd plugins/woocommerce && USE_WP_ENV=1 pnpm playwright test --config=e2e/playwright.config.js ./e2e/tests/activate-and-setup/basic-setup.spec.js (running a single test)
```
Please, let us know if anything is not clear or if you need more information!

**Comment by tomalec** [comment]
Sorry, @alopezari but I'm still pretty much confused.

> So yes, you can rely on these Playwright specs to test this PR :)

What you suggest [above](https://github.com/woocommerce/woocommerce/pull/33828#issuecomment-1211793226) is to "test" - check that this PR is acceptable? "test" - check that this PR solves the original issue, which was running locally e2e tests in the extension I develop? or "test" in what sense?

> You can find [here](https://github.com/woocommerce/woocommerce/blob/db50c3249720711f61f7bae29b8a7b5ffba53ba2/plugins/woocommerce/e2e/README.md#introduction) more information about how to run these tests. In concrete, you can use this example to run individually the specs mentioned above, which covers the same as the previous Admin E2E tests:

Butt this is about testing WooCommerce - the plugin from this repo, and not my use case which is testing WooCommerce **extension** locally.

**Comment by tomalec** [comment]
Also, speaking of how to run all of that in this repo. `pnpm install` fails, as it requires an old version of `pnpm` to which I cannot downgrade:
```
$ nvm use && pnpm install

Found '/home/tomalec/repos/woocommerce/.nvmrc' with version <v16>
Now using node v16.14.2 (npm v8.5.0)
 ERR_PNPM_UNSUPPORTED_ENGINE  Unsupported environment (bad pnpm and/or Node.js version)

Your pnpm version is incompatible with "/home/tomalec/repos/woocommerce/plugins/woocommerce".

Expected version: ^6.24.2
Got: 7.5.2

This is happening because the package's manifest has an engines.pnpm field specified.
To fix this issue, install the required pnpm version globally.

To install the latest version of pnpm, run "pnpm i -g pnpm".
To check your pnpm version, run "pnpm -v".

$ pnpm i -g pnpm@6.24.2
Packages: +1
+
Packages are hard linked from the content-addressable store to the virtual store.
  Content-addressable store is at: /home/tomalec/.local/share/pnpm/store/v3
  Virtual store is at:             .pnpm

/home/tomalec/.local/share/pnpm/global/5:
+ pnpm 6.24.2 (7.9.0 is available)

Progress: resolved 1, reused 0, downloaded 1, added 1, done

$ pnpm i
 ERR_PNPM_UNSUPPORTED_ENGINE  Unsupported environment (bad pnpm and/or Node.js version)

Your pnpm version is incompatible with "/home/tomalec/repos/woocommerce/plugins/woocommerce".

Expected version: ^6.24.2
Got: 7.5.2

This is happening because the package's manifest has an engines.pnpm field specified.
To fix this issue, install the required pnpm version globally.

To install the latest version of pnpm, run "pnpm i -g pnpm".
To check your pnpm version, run "pnpm -v".

```

**Comment by alopezari** [comment]
> What you suggest [above](https://github.com/woocommerce/woocommerce/pull/33828#issuecomment-1211793226) is to "test" - check that this PR is acceptable? "test" - check that this PR solves the original issue, which was running locally e2e tests in the extension I develop? or "test" in what sense?
> Butt this is about testing WooCommerce - the plugin from this repo, and not my use case which is testing WooCommerce **extension** locally.

Sorry @tomalec, I missed the extension part and thought you wanted to run regression tests to check the config change didn't cause any major inconveniences. Sorry for the confusion.

To be honest, I don't have a clear answer now about how to proceed in your case (this applies also to the PNPM version problem), since we're in the middle of a migration and some things are changing. But I will help you unblock this PR considering it's been here for a while. I'll look for the answers and come back to you once I get something.

**Comment by tomalec** [comment]
Through my manual cop'n'pasting & checks mentioned in [OP](https://github.com/woocommerce/woocommerce/pull/33828#issue-1302409247), I managed to test it solves at least one step with setting up admin e2e tests for external extension.

Speaking of running local `woocommerce/woocommerce` builds and checks, I still struggle to run `pnpm run build`. I managed to downgrade my `pnpm` to `6.24`, but what I got is:
```
 ~/repos/woocommerce   trunk  nvm use
Found '/home/tomalec/repos/woocommerce/.nvmrc' with version <v16>
Now using node v16.14.2 (npm v8.5.0)
  ~/repos/woocommerce   trunk  pnpm --version
6.24.2
  ~/repos/woocommerce   trunk  pnpm run build

> woocommerce-monorepo@ build /home/tomalec/repos/woocommerce
> pnpm exec turbo run turbo:build -- -- 

 ERROR  workspace configuration error: package.json: no workspaces found. Turborepo requires npm workspaces to be defined in the root package.json
 ELIFECYCLE  Command failed with exit code 1.
```

**Comment by zhongruige** [comment]
Hi @tomalec, thanks for getting this in! Looks like we have a merge conflict on `pnpm-lock.yaml` that we'll need to resolve before we can merge in these changes.

**Comment by rodelgc** [comment]
@tomalec once the conflicts are resolved, feel free to ping me for a re-review so we could merge it in. Thanks!

**Comment by tomalec** [comment]
@rodelgc I redid the changes as the conflicts were too far behind. I also added `config` to syncpack to keep it on track in the future

**Comment by alopezari** [comment]
Hi @tomalec!

The changes look good to me.

I followed the testing instructions with the [google-listings-and-ads](https://github.com/woocommerce/google-listings-and-ads/tree/b14771982bb0c84d6252b27f36684114da06b85a) plugin and the tests failed always with the same failure:
```
At setup page › Merchant who has their Jetpack connected, but Google not yet connected › after clicking the "Connect your Google account" button › should sent an API request to connect Google account, and redirect to the returned URL

    expect(received).resolves.toMatch(expected)

    Expected substring: "Log In"
    Received string:    "WordPress › Installation"
```

However, I don't think it's related to the config change since these tests also fail for that git reference (b14771982bb0c84d6252b27f36684114da06b85a) without updating to config@3.3.7 (in other words, without the changes from this PR). It looks good to me since there isn't any execution error.

The config version was also updated in the WooCommerce `package.json`, and the tests passed in this PR, so that looks good to me too.

However, considering the WooCommerce `package.json` has been updated, I'd like to add @woocommerce/atlas to the review to double check, and also because they are the Monorepo owners.

Apart from the review from Atlas, there are new conflicts in the `pnpm-lock.yml`. Once it is solved and Atlas approves the review, we're happy to merge this 👍 Thank you for your work and patience!

**Comment by tomalec** [comment]
I rebased and force-pushed changes to solve pnpm conflicts.

> I followed the testing instructions with the [google-listings-and-ads](https://github.com/woocommerce/google-listings-and-ads/tree/b14771982bb0c84d6252b27f36684114da06b85a) plugin and the tests failed always with the same failure:

That branch was made over 6months ago, I guess the entire testing environment changed a lot. Given it was not implemented on a stable foundation back then, and there was still not a closed set of problems with importing `admin-e2e-tests`, I don't expect it to run 100% successfully today.

File: .syncpackrc
Comment (samueljseay) [inline .syncpackrc:38]: In syncpack the `filter` attribute means that syncpack will ignore any dependencies that don't match the filter even if they have a version group object like this.

The API is sort of counterintuitive for what we want to achieve right now and there are possibly some alternatives, but for now you'll need to add "config" to the filter regex if you want this to work.

**Review by samueljseay — CHANGES_REQUESTED** [review]

File: .syncpackrc
Comment (tomalec) [reply .syncpackrc:38]: Thanks, I missed it.

Addressed in https://github.com/woocommerce/woocommerce/pull/33828/commits/105d95ab6d1b56c8896320d744016db650a242ec

> The API is sort of counterintuitive for what we want to achieve right now

On the other way around: API is intuitive, but our usage is against the design. AFAIU, syncpack by default, tries to keep all packages in sync, then expose `versionGroups` to group the exception rules.

What we do is, ignore everything and sync only a few. This is probably the only way to start cleaning it up gradually, without "do it all" revolution.

**Comment by tomalec** [comment]
@samueljseay I resolved conflicts and the PR is ready for another look :)

**Review by samueljseay — DISMISSED** [review]
Looking good 🚀

**Review by nigeljamesstevenson — DISMISSED** [review]
Hi folks, I am Porter this week and it appears this one has slipped through the net with the Woo DM last week...

Unfortunately, there are some conflicts:

- .syncpackrc
- pnpm-lock.yaml

Feel free to resolve and I will merge - @samueljseay has already stated that he is happy with this one - am sure @tomalec would like to get this one done and dusted 😊

**Comment by tomalec** [comment]
Hi @nigeljamesstevenson, I resolved the conflicts again. Could you approve and merge?

**Review by samueljseay — APPROVED** [review]
🚀 Thank you so much for this, love seeing everyone help us move forward with syncpack.

**Pull request merged**
